File: juddi-client/src/main/java/org/apache/juddi/v3/client/util/TModelInstanceDetailsComparator.java
/*
 * Copyright 2013 The Apache Software Foundation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.juddi.v3.client.util;

import javax.xml.datatype.DatatypeConfigurationException;
import javax.xml.datatype.DatatypeConstants;
import javax.xml.datatype.DatatypeFactory;
import javax.xml.datatype.Duration;
import javax.xml.datatype.XMLGregorianCalendar;
import org.uddi.api_v3.InstanceDetails;
import org.uddi.api_v3.TModelInstanceDetails;

/**
 * Compares two UDDI TModelInstanceDetails, searching for a specific tModel key,
 * then parsing to the selected data type, then comparing.
 *
 * @author Alex O'Ree
 */
public class TModelInstanceDetailsComparator {

    /**
     *
     * @param TModelKey for TModelInstanceInfo to use for comparison
     * @param number if true, the InstanceDetails.InstanceParms will be treated
     * like a number
     * @param XMLdate if true, the InstanceDetails.InstanceParms will be treated
     * like a XML Date
     * @param XMLduration if true, the InstanceDetails.InstanceParms will be
     * treated like a XML Gregorian Calendar
     * @see Duration
     * @see XMLGregorianCalendar
     * @throws DatatypeConfigurationException
     */
    public TModelInstanceDetailsComparator(String TModelKey, boolean number, boolean XMLdate, boolean XMLduration) throws DatatypeConfigurationException, IllegalArgumentException {
        if (TModelKey == null || TModelKey.length() == 0) {
            throw new IllegalArgumentException();
        }
        compareField = TModelKey;
        if (!number && !XMLdate && !XMLduration) {
            throw new IllegalArgumentException("only one data type can be selected");
        }
        if (number && XMLdate && !XMLduration) {
            throw new IllegalArgumentException("only one data type can be selected");
        }
        if (number && !XMLdate && XMLduration) {
            throw new IllegalArgumentException("only one data type can be selected");
        }
        if (!number && XMLdate && XMLduration) {
            throw new IllegalArgumentException("only one data type can be selected");
        }
        if (number && XMLdate && XMLduration) {
            throw new IllegalArgumentException("only one data type can be selected");
        }
        fac = DatatypeFactory.newInstance();
        isNumber = number;
        isDate = XMLdate;
        isDuration = XMLduration;
    }
    DatatypeFactory fac = null;
    String compareField = null;
    boolean isNumber = false;
    boolean isDate = false;
    boolean isDuration = false;

    /**
     * Compares two non-null instances of TModelInstanceDetails by only
     * comparing the field designated from the constructor. It will also cast or
     * parse TModelInstanceDetails[i].InstanceDetails[k].InstanceParms to the
     * selected data type double, XMLGregorgian or Duration, using that as a
     * basis for comparison. If a parsing error occurs, an exception will be
     * thrown.
     *
     * @param lhs
     * @param rhs
     * @return less than 0 if lhs &lt; rhs, greater than 0 if lhs &gt; rhs.
     * @throws IllegalArgumentException if the tModel key to search for is
     * missing, if either sides are null
     * @throws ArrayIndexOutOfBoundsException if the values were found but could
     * not be compared
     */
    public int compare(TModelInstanceDetails lhs, TModelInstanceDetails rhs) throws IllegalArgumentException, NumberFormatException, NullPointerException, ArrayIndexOutOfBoundsException {
        if (lhs == null) {
            throw new IllegalArgumentException("lhs");
        }
        if (rhs == null) {
            throw new IllegalArgumentException("rhs");
        }
        if (lhs.getTModelInstanceInfo().isEmpty() || rhs.getTModelInstanceInfo().isEmpty()) {
            throw new IllegalArgumentException("no data to compare");
        }
        InstanceDetails lhsc = null;
        InstanceDetails rhsc = null;
        for (int i = 0; i < lhs.getTModelInstanceInfo().size(); i++) {
            if (lhs.getTModelInstanceInfo().get(i).getTModelKey().equalsIgnoreCase(compareField)) {
                lhsc = lhs.getTModelInstanceInfo().get(i).getInstanceDetails();
            }
        }
        for (int i = 0; i < rhs.getTModelInstanceInfo().size(); i++) {
            if (rhs.getTModelInstanceInfo().get(i).getTModelKey().equalsIgnoreCase(compareField)) {
                rhsc = rhs.getTModelInstanceInfo().get(i).getInstanceDetails();
            }
        }

        if (lhsc == null) {
            throw new IllegalArgumentException(compareField + " not found for lhs");
        }
        if (rhsc == null) {
            throw new IllegalArgumentException(compareField + " not found for rhs");
        }
        if (lhsc.getInstanceParms() == null) {
            throw new IllegalArgumentException(compareField + " found lhs, but no data");
        }
        if (rhsc.getInstanceParms() == null) {
            throw new IllegalArgumentException(compareField + " found rhs, but no data");
        }
        if (isNumber) {
            Double l = Double.parseDouble(lhsc.getInstanceParms());
            Double r = Double.parseDouble(rhsc.getInstanceParms());
            return l.compareTo(r);
        }

        if (isDate) {
            XMLGregorianCalendar l = fac.newXMLGregorianCalendar(lhsc.getInstanceParms());
            XMLGregorianCalendar r = fac.newXMLGregorianCalendar(rhsc.getInstanceParms());
            //System.out.println(l.toXMLFormat() + " " + r.toXMLFormat());
            int x = l.compare(r);

            if (x == DatatypeConstants.LESSER) {
                return -1;
            }

            if (x == DatatypeConstants.GREATER) {
                return 1;
            }
            if (x == DatatypeConstants.EQUAL) {
                return 0;
            }
            throw new ArrayIndexOutOfBoundsException("cannot compare, result was " + x);
        }

        if (isDuration) {
            Duration l = fac.newDuration(lhsc.getInstanceParms());
            Duration r = fac.newDuration(rhsc.getInstanceParms());
          //  System.out.println(l.toString() + " " + r.toString());
            int x = l.compare(r);

            if (x == DatatypeConstants.LESSER) {
                return -1;
            }

            if (x == DatatypeConstants.GREATER) {
                return 1;
            }
            if (x == DatatypeConstants.EQUAL) {
                return 0;
            }
            throw new ArrayIndexOutOfBoundsException("cannot compare, result was " + x);
        }

        return 0;
    }
}

File: juddi-client.net/juddi-client.net/org.apache.juddi.v3.client.util/TModelInstanceDetailsComparator.cs
using org.uddi.apiv3;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

namespace org.apache.juddi.v3.client.util
{
    /// <summary>
    /// Compares two UDDI TModelInstanceDetails, searching for a specific tModel key, then parsing to the selected data type, then comparing.
    /// @author Alex O'Ree
    /// </summary>
    public class TModelInstanceDetailsComparator
    {

        /**
         *
         * @param TModelKey for TModelInstanceInfo to use for comparison
         * @param number if true, the InstanceDetails.InstanceParms will be treated
         * like a number
         * @param XMLdate if true, the InstanceDetails.InstanceParms will be treated
         * like a XML Date
         * @param XMLduration if true, the InstanceDetails.InstanceParms will be
         * treated like a XML Gregorian Calendar
         * @see Duration
         * @see XMLGregorianCalendar
         * @throws DatatypeConfigurationException, ArgumentOutOfRangeException, ArgumentNullException
         * FormatException
         */
        public TModelInstanceDetailsComparator(String TModelKey, bool number, bool XMLdate, bool XMLduration)
        {
            if (TModelKey == null || TModelKey.Length == 0)
            {
                throw new ArgumentNullException();
            }
            compareField = TModelKey;
            if (!number && !XMLdate && !XMLduration)
            {
                throw new ArgumentOutOfRangeException("only one data type can be selected");
            }
            if (number && XMLdate && !XMLduration)
            {
                throw new ArgumentOutOfRangeException("only one data type can be selected");
            }
            if (number && !XMLdate && XMLduration)
            {
                throw new ArgumentOutOfRangeException("only one data type can be selected");
            }
            if (!number && XMLdate && XMLduration)
            {
                throw new ArgumentOutOfRangeException("only one data type can be selected");
            }
            if (number && XMLdate && XMLduration)
            {
                throw new ArgumentOutOfRangeException("only one data type can be selected");
            }
            isNumber = number;
            isDate = XMLdate;
            isDuration = XMLduration;
        }

        String compareField = null;
        bool isNumber = false;
        bool isDate = false;
        bool isDuration = false;

        /**
         * Compares two non-null instances of TModelInstanceDetails by only
         * comparing the field designated from the constructor. It will also cast or
         * parse TModelInstanceDetails[i].InstanceDetails[k].InstanceParms to the
         * selected data type double, XMLGregorgian or Duration, using that as a
         * basis for comparison. If a parsing error occurs, an exception will be
         * thrown.
         *
         * @param lhs
         * @param rhs
         * @return less than 0 if lhs &lt; rhs, greater than 0 if lhs &gt; rhs.
         * @throws IllegalArgumentException if the tModel key to search for is
         * missing, if either sides are null
         * @throws ArrayIndexOutOfBoundsException if the values were found but could
         * not be compared
         *  throws IllegalArgumentException, NumberFormatException, NullPointerException, ArrayIndexOutOfBoundsException
         */
        public int compare(tModelInstanceInfo[] lhs, tModelInstanceInfo[] rhs)
        {
            if (lhs == null)
            {
                throw new ArgumentNullException("lhs");
            }
            if (rhs == null)
            {
                throw new ArgumentNullException("rhs");
            }
            if (lhs.Length == 0 || rhs.Length == 0)
            {
                throw new ArgumentOutOfRangeException("no data to compare");
            }
            instanceDetails lhsc = null;
            instanceDetails rhsc = null;
            for (int i = 0; i < lhs.Length; i++)
            {
                if (lhs[i].tModelKey!=null && lhs[i].tModelKey.Equals(compareField, StringComparison.CurrentCultureIgnoreCase))
                {
                    lhsc = lhs[i].instanceDetails;
                }
            }
            for (int i = 0; i < rhs.Length; i++)
            {
                if (rhs[i].tModelKey!=null && rhs[i].tModelKey.Equals(compareField, StringComparison.CurrentCultureIgnoreCase))
                {
                    rhsc = rhs[i].instanceDetails;
                }
            }

            if (lhsc == null)
            {
                throw new ArgumentOutOfRangeException(compareField + " not found for lhs");
            }
            if (rhsc == null)
            {
                throw new ArgumentOutOfRangeException(compareField + " not found for rhs");
            }
            if (lhsc.instanceParms == null)
            {
                throw new ArgumentOutOfRangeException(compareField + " found lhs, but no data");
            }
            if (rhsc.instanceParms == null)
            {
                throw new ArgumentOutOfRangeException(compareField + " found rhs, but no data");
            }
            if (isNumber)
            {
                Double l = Double.Parse(lhsc.instanceParms);
                Double r = Double.Parse(rhsc.instanceParms);
                return l.CompareTo(r);
            }

            if (isDate)
            {

                DateTime l = XmlConvert.ToDateTime(lhsc.instanceParms);
                DateTime r = XmlConvert.ToDateTime(rhsc.instanceParms);
                int x = l.CompareTo(r);

                return x;
            }

            if (isDuration)
            {
                TimeSpan l = XmlConvert.ToTimeSpan(lhsc.instanceParms);
                TimeSpan r = XmlConvert.ToTimeSpan(rhsc.instanceParms);
                int x = l.CompareTo(r);
                return x;
            }

            return 0;
        }
    }
}
